java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationState.java

```diff
@@ -31,30 +29,38 @@
   /**
    * This value means that the checked relation is fulfilled by the set of known relations
    */
-  FULFILLED(true, BooleanConstraint.FALSE),
+  FULFILLED,
   /**
    * This value means that the checked relation is not fulfilled by the set of known relations
    */
-  UNFULFILLED(true, BooleanConstraint.TRUE),
+  UNFULFILLED,
   /**
    * This value means that the checked relation is not determined by the set of known relations
    */
-  UNDETERMINED(false, null);
-
-  private final boolean determined;
-  private final BooleanConstraint checkedConstraint;
-
-  RelationState(boolean determined, @Nullable BooleanConstraint checkedConstraint) {
-    this.determined = determined;
-    this.checkedConstraint = checkedConstraint;
-  }
+  UNDETERMINED;
 
   public boolean rejects(BooleanConstraint constraint) {
-    return checkedConstraint != null && checkedConstraint.equals(constraint);
+    if (this == FULFILLED) {
+      return constraint == BooleanConstraint.FALSE;
+    }
+    if (this == UNFULFILLED) {
```

Comment:
intellij is suggesting to replace this with : 
``` 
 return this == UNFULFILLED && constraint == BooleanConstraint.TRUE;
```

Not sure if it is much better, up to you.